### PR TITLE
Fix S1144 FP: types used through reflection

### LINKED ISSUE #9379: Fix S1144 FP: Diagnostic doesn't respect reflection with `DynamicallyAccessedMembers` attribute
### Description

Diagnostic S1144 doesn't respect reflection even though the generic parameter is annotated with the `DynamicallyAccessedMembers` attribute.

Related: #7774

### Repro steps

This is a minified repro. In my actuall use case I called [`UseMiddleware`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.usemiddlewareextensions.usemiddleware?view=aspnetcore-8.0).

```csharp
using System.Diagnostics.CodeAnalysis;

namespace SonarCubeRepro
{
    internal static class Program
    {
        static void Main()
        {
            var instance = CreateInstance<Foo>();
            Console.WriteLine( instance.Number );
        }

        public static T CreateInstance<[DynamicallyAccessedMembers( DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors )] T>()
        {
            var instance = (T)Activator.CreateInstance( typeof( T ), 42 )!;
            return instance;
        }

        private sealed class Foo
        {
            public Foo( int number )
            {
                Number = number;
            }

            public int Number { get; set; }
        }
    }
}
```

### Expected behavior

No warning.

### Actual behavior

Warning S1144

### Known workarounds

Supress the issue

### Related information

SonarAnalyzer.CSharp Version=9.26.0.92422

* C#/VB.NET Plugins version
* Visual Studio version: 17.9.6
* MSBuild / dotnet version: 8.0.204
* SonarScanner for .NET version (if used)
* Operating System: Windows 10

### PR DESCRIPTION
**Pull request by zsolt-kolbay-sonarsource**
Fixes #9379

RSpec change: https://github.com/SonarSource/rspec/pull/3992

### CODE REVIEW

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.Test/TestCases/UnusedPrivateMember.CSharp9.cs [hunk 1]
```diff
@@ -168,3 +169,100 @@ public class ClassWithPrintMembers
...
+
+            private class NestedType { }                                        // Noncompliant
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.NonPublicNestedTypes)] private class NestedTypeWithAttribute { }
+        }
+
+        private class ArgumentsDecoratedWithAttribute   // Noncompliant
+        {
+            public void PublicMethod() => Method(null); // Noncompliant
+
+            private void Method([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] NestedType arg) { }
```
[the remark this answers is not in the document]
Comment (CristianAmbrosini) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/UnusedPrivateMember.CSharp9.cs:248]: There's already a test for private properties

File: analyzers/src/SonarAnalyzer.CSharp/Rules/UnusedPrivateMember.cs [hunk 2]
```diff
@@ -151,6 +151,20 @@ private static void NamedSymbolAction(SonarSymbolReportingContext context, HashS
         ReportDiagnosticsForMembers(context, unusedSymbols, accessibility, fieldLikeSymbols);
     }
 
+    private static bool IsUsedWithReflection(ISymbol symbol, HashSet<ISymbol> symbolsUsedWithReflection)
+    {
+        var currentSymbol = symbol;
+        while (currentSymbol is not null)
+        {
+            if (symbolsUsedWithReflection.Contains(currentSymbol))
```
[the remark this answers is not in the document]
Comment (CristianAmbrosini) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/UnusedPrivateMember.cs:159]: This is compliant with the current implentation, I'm not sure we need the OriginalDefinition

File: analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs [hunk 3]
```diff
@@ -134,19 +135,26 @@ static int GetTupleCount(ExpressionSyntax assignmentLeft)
 
         public override void VisitAttribute(AttributeSyntax node)
         {
-            var symbol = semanticModel.GetSymbolInfo(node).Symbol;
-            if (symbol != null
-                && symbol.ContainingType.Is(KnownType.System_Diagnostics_DebuggerDisplayAttribute)
-                && node.ArgumentList != null)
+            if (semanticModel.GetSymbolInfo(node).Symbol is { } symbol)
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs:138]: NIT: as in the other PR: `semanticModel.GetSymbolInfo(node).Symbol is IMethodSymbol { MethodKind:Constructor, Parent: ITypeSymbol attribute } `

File: analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs [hunk 4]
```diff
@@ -169,6 +177,27 @@ public override void VisitIdentifierName(IdentifierNameSyntax node)
...
+            {
+                var typesWithDynamicallyAccessedMembers = GetMethodTypeParameters(symbol)
+                    .Zip(node.Arguments, (symbol, argument) => new Tuple<ITypeParameterSymbol, SyntaxNode>(symbol, argument)) // map T to Person in void M<T>() { } .. M<Person>();
+                    .Where(x => x.Item1.HasAttribute(KnownType.System_Diagnostics_CodeAnalysis_DynamicallyAccessedMembersAttribute))
+                    .Select(x => semanticModel.GetSymbolInfo(x.Item2).Symbol);
+                TypesUsedWithReflection.UnionWith(typesWithDynamicallyAccessedMembers);
+            }
+            base.VisitTypeArgumentList(node);
+
+            static ImmutableArray<ITypeParameterSymbol> GetMethodTypeParameters(ISymbol symbol) =>
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs:192]: Might be inlined or needs to be named `GetTypeParameters`

File: analyzers/tests/SonarAnalyzer.Test/TestCases/UnusedPrivateMember.CSharp9.cs [hunk 5]
```diff
@@ -168,3 +169,121 @@ public class ClassWithPrintMembers
...
+        public static void Method()
+        {
+            var instance = CreateInstance<ClassInstantiatedThroughReflection>();
+
+            A classViaReflection = new();
+            InitValue(classViaReflection); 
+        }
+
+        public static T CreateInstance<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] T>() =>
+            (T)Activator.CreateInstance(typeof(T), 42);
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/UnusedPrivateMember.CSharp9.cs:187]: According [to the docs](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.dynamicallyaccessedmembersattribute?view=net-8.0#remarks), the attribute also works for `string` and `Type` parameters:
```cs
public static object CreateInstance([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type type) => // ..

CreateInstance(typeof(SomePrivateClass));
```

You may want to add an FP UT.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/UnusedPrivateMember.CSharp9.cs [hunk 6]
```diff
@@ -168,3 +169,121 @@ public class ClassWithPrintMembers
...
+
+        public static void InitValue<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] T>(T value) { }
+
+        private class A
+        {
+            private bool a = true; // Noncompliant FP: type argument is inferred and ArgumentList is not present on syntax level
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/UnusedPrivateMember.CSharp9.cs:193]: Can you point to the place where the FP originates from. It is hard to see that line 183 is related to this code.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/UnusedPrivateMember.CSharp9.cs [hunk 7]
```diff
@@ -168,3 +169,121 @@ public class ClassWithPrintMembers
...
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.NonPublicProperties)] private int PrivatePropertyWithAttribute { get; set; }
+
+            private void PrivateMethod() { }                                    // Noncompliant
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.NonPublicMethods)] private void PrivateMethodWithAttribute() { }
+
+            private class NestedType { }                                        // Noncompliant
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.NonPublicNestedTypes)] private class NestedTypeWithAttribute { }
+
+            [return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.NonPublicMethods)]
+            private int PrivateMethodWithReturn() { return 0; } // Noncompliant
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/UnusedPrivateMember.CSharp9.cs:260]: The FP here is that the return type is not detected. Replace `int` with some private class a trimmer might remove.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
It generally looks fine. It misses some UTs that might be worth adding. Feel free not to add them if you think it's not worth it.

**Comment by zsolt-kolbay-sonarsource** [comment]
I removed the check through type parameters, because it nearly doubled the analysis time.
We'll need to accept it as FP.

**Comment by martin-strecker-sonarsource** [comment]
> I removed the check through type parameters, because it nearly doubled the analysis time.
We'll need to accept it as FP.

How did you measure? If the rule was very, very fast, duplication is not a problem. Can you post some more info about the perf validation?

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

**Comment by zsolt-kolbay-sonarsource** [comment]
> > I removed the check through type parameters, because it nearly doubled the analysis time.
> > We'll need to accept it as FP.
> 
> How did you measure? If the rule was very, very fast, duplication is not a problem. Can you post some more info about the perf validation?

I used the [AnalyzerRunner](https://github.com/dotnet/roslyn/tree/main/src/Tools/AnalyzerRunner) from the Roslyn repo to compare the performance with and without the check for type parameters. I ran it on _AutoMapper_, _Akka.Net_, and an earlier version of the _Roslyn_ repo. They all showed a degradation in performance ranging between 40% and 90%. `UnusedPrivateMember` is one of the slowest rules we have (excluding SE rules), so I don't think such a change is acceptable.

```
WITHOUT the check for type arguments (AutoMapper):

Statistics for UnusedPrivateMember:
Concurrent:                     False
Execution time (ms):            3644.8062
Code Block Actions:             0
Code Block Start Actions:       0
Code Block End Actions:         0
Compilation Actions:            0
Compilation Start Actions:      6
Compilation End Actions:        6
Operation Actions:              0
Operation Block Actions:        0
Operation Block Start Actions:  0
Operation Block End Actions:    0
Semantic Model Actions:         0
Symbol Actions:                 6
Symbol Start Actions:           0
Symbol End Actions:             0
Syntax Node Actions:            0
Syntax Tree Actions:            0
Additional File Actions:        0
Suppression Actions:            0
Execution times (ms):
UnusedPrivateMember:    3645
  S1144: 1364 instances
  S4487: 9 instances

--------------------------------------------------------
--------------------------------------------------------

INCLUDING the check for type arguments (AutoMapper):

Statistics for UnusedPrivateMember:
Concurrent:                     False
Execution time (ms):            5731.9556
Code Block Actions:             0
Code Block Start Actions:       0
Code Block End Actions:         0
Compilation Actions:            0
Compilation Start Actions:      6
Compilation End Actions:        6
Operation Actions:              0
Operation Block Actions:        0
Operation Block Start Actions:  0
Operation Block End Actions:    0
Semantic Model Actions:         0
Symbol Actions:                 6
Symbol Start Actions:           0
Symbol End Actions:             0
Syntax Node Actions:            0
Syntax Tree Actions:            0
Additional File Actions:        0
Suppression Actions:            0
Execution times (ms):
UnusedPrivateMember:    5732
  S1144: 1364 instances
  S4487: 9 instances
```

File: analyzers/packaging/output/Akka.API.Tests/net471/SonarProjectConfig.xml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<SonarProjectConfig xmlns="http://www.sonarsource.com/msbuild/analyzer/2021/1">
```
Comment (martin-strecker-sonarsource) [inline analyzers/packaging/output/Akka.API.Tests/net471/SonarProjectConfig.xml:1]: Upps. These files do not belong here.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Can you also update the RSpec? It needs the `DynamicallyAccessedMembersAttribute` in the exception section. Consider only checking only the name of the attribute instead of namespace/name. This would allow users to create this attribute in unsupported platforms (which is every platform except Net5 upwards). The name is long enough to not be confused with anything else. If you go that route, tests are needed.

File: analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs [hunk 9]
```diff
@@ -18,415 +18,427 @@
...
-                || a.NameColon.Name.Identifier.ValueText == "Type";
-        }
+        static ISymbol FindDeconstructor(IEnumerable<ISymbol> deconstructors, int numberOfArguments) =>
+            deconstructors.FirstOrDefault(m => m.GetParameters().Count() == numberOfArguments && m.DeclaredAccessibility.IsAccessibleOutsideTheType());
+    }
 
-        public override void VisitIdentifierName(IdentifierNameSyntax node)
+    public override void VisitAttribute(AttributeSyntax node)
+    {
+        if (node.GetName() is "DynamicallyAccessedMembersAttribute" or "DynamicallyAccessedMembers"
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs:139]:
```suggestion
        // Some members that seem unused might be dynamically accessed through reflection.
        // The DynamicallyAccessedMembersAttribute was introduced to inform tools about such uses.
        // The attribute is not available on NetFramework and we want to enable this mechanism for these users. 
        // Therefore, we only check the name, but not the namespace and let the users define their own custom version.
        // https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.dynamicallyaccessedmembersattribute
        if (node.GetName() is "DynamicallyAccessedMembersAttribute" or "DynamicallyAccessedMembers"
```

Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs:139]: Consider making "DynamicallyAccessedMembers" a const:
```suggestion
        const string DynamicallyAccessedMembers = "DynamicallyAccessedMembers";
        if (node.GetName() is DynamicallyAccessedMembers or $"{DynamicallyAccessedMembers}Attribute"
```

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM. I think a comment is helpful in this case.

**Pull request merged**